### PR TITLE
Added session to smoothen user experience - see readme for details

### PR DESCRIPTION
**Pull request by aremo-ms**
When a ChallengeUser method is called, the redirect URL is pointing to GET method by default. When the challenge happen during Create POST, for example, user will have a bad experience after being redirected to Create GET. One of the ways to smoothen user experience is to use browser cookies to store fields information.

### CODE REVIEW

File: TodoListClient/Controllers/TodoListController.cs [hunk 1]
```diff
@@ -47,6 +43,9 @@ public Todo Get(int id)
         [AuthorizeForScopes(ScopeKeySection = "TodoList:TodoListScope")]
         public ActionResult Index()
         {
+            //reset cookies on every entry to TODO's list
```
Comment (kalyankrishna1) [inline TodoListClient/Controllers/TodoListController.cs]: >cookies [](http://example.com/codeflow?start=20&length=7)

session

File: TodoListClient/Controllers/TodoListController.cs [hunk 2]
```diff
@@ -64,9 +63,21 @@ public ActionResult Create()
...
                 _consentHandler.ChallengeUser(new string[] { "user.read" }, claimsChallenge);
+
                 return new EmptyResult();
             }
 
+            var todoObject = TodoSessionState(SessionAction.Get);
+
+            if (todoObject != null && todoObject.IsInitialized())
+            {
+                StoreTodo(todoObject);
```
Comment (kalyankrishna1) [inline TodoListClient/Controllers/TodoListController.cs]: >StoreTodo [](http://example.com/codeflow?start=16&length=9)

StoreToDoToDB

File: TodoListClient/Models/TodoItem.cs [hunk 3]
```diff
@@ -12,5 +7,10 @@ public class Todo
         public string Title { get; set; }
 
         public string Owner { get; set; }
+
+        internal bool IsInitialized()
```
Comment (kalyankrishna1) [inline TodoListClient/Models/TodoItem.cs]: >IsInitialized [](http://example.com/codeflow?start=22&length=13)

One line comment would be nice as its not as clear in first look as what it is, unlike other attributes

File: README.md [hunk 4]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
 
 ```
 
+### Using Session State
+
+When a ChallengeUser method is called, the redirect URL is pointing to GET method by default. When the challenge happen during Create POST, for example, user will have a bad experience after being redirected to Create GET. One of the ways to smoothen user experience is to use session state to store fields information.
```
Comment (kalyankrishna1) [inline README.md:392]: >a [](http://example.com/codeflow?start=5&length=1)

the

File: README.md [hunk 5]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
 
 ```
 
+### Using Session State
```
Comment (kalyankrishna1) [inline README.md]: >Using Session State [](http://example.com/codeflow?start=4&length=19)

Using Session state for a better user experience

File: README.md [hunk 6]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
...
+    public ActionResult Create()
+    {
+        string claimsChallenge = CheckForRequiredAuthContext(Request.Method);
+
+        if (!string.IsNullOrWhiteSpace(claimsChallenge))
+        {
+            _consentHandler.ChallengeUser(new string[] { "user.read" }, claimsChallenge);
+
+            return new EmptyResult();
+        }
```
Comment (kalyankrishna1) [inline README.md:405]: change it to 
.... some code

File: README.md [hunk 7]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
...
+        if (todo != null && todoObject.IsInitialized())
+        {
+            StoreTodo(todoObject);
+
+            TodoSessionState(SessionAction.Set);
+
+            return RedirectToAction("Index");
+        }
+
+        Todo todo = new Todo() { Owner = HttpContext.User.Identity.Name };
```
Comment (kalyankrishna1) [inline README.md:413]: change it to 
.... some more code

File: README.md [hunk 8]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
...
+    // POST: TodoList/Create
+    [HttpPost]
+    [ValidateAntiForgeryToken]
+    public ActionResult Create([Bind("Title,Owner")] Todo todo)
+    {
+        string claimsChallenge = CheckForRequiredAuthContext(Request.Method);
+
+        if (!string.IsNullOrWhiteSpace(claimsChallenge))
+        {
+            _consentHandler.ChallengeUser(new string[] { "user.read" }, claimsChallenge);
```
Comment (kalyankrishna1) [inline README.md:424]: change it to 
.... some code

File: README.md [hunk 9]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
...
+
+            TodoSessionState(SessionAction.Set, todo);
+
+            return new EmptyResult();
+        }
+
+        StoreTodo(new Todo() { Owner = HttpContext.User.Identity.Name, Title = todo.Title });
```
Comment (kalyankrishna1) [inline README.md]: change it to 
.... some more code

File: README.md [hunk 10]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
...
+                    break;
+            }
+
+            return todo;
+    }
+    private enum SessionAction
+    {
+        Set,
+        Get
+    }
```
Comment (kalyankrishna1) [inline README.md]: you can omit this if you wish to keep the readme small

File: README.md [hunk 11]
```diff
@@ -387,6 +387,90 @@ if (!string.IsNullOrWhiteSpace(claimsChallenge))
 
 ```
 
+### Using Session State
+
+When a ChallengeUser method is called, the redirect URL is pointing to GET method by default. When the challenge happen during Create POST, for example, user will have a bad experience after being redirected to Create GET. One of the ways to smoothen user experience is to use session state to store fields information.
+Take a look into the example of using session state.
```
Comment (kalyankrishna1) [inline README.md:394]: When the `ChallengeUser()` method is called, the redirect URL of the request sent to Azure AD will bring the user back to the /GET path. So the user might get confused if the authentication context was applied to an action like /CREATE. To make for a smoother user experience, we use the [ASP.NET Session state](https://docs.microsoft.com/aspnet/core/fundamentals/app-state) to store the user's input. This allows us to restore the ToDo item once the user has come back to the web app after redirection.

**Review by kalyankrishna1 — APPROVED** [review]
:shipit:

File: TodoListClient/Controllers/TodoListController.cs (hunk 2 above)
Comment (aremo-ms) [reply TodoListClient/Controllers/TodoListController.cs]: I suggest to call it PersistToDo

File: README.md (hunk 10 above)
Comment (aremo-ms) [reply README.md]: done

File: README.md (hunk 9 above)
Comment (aremo-ms) [reply README.md]: @kalyankrishna1 
Please clarify what you mean here

**Review by kalyankrishna1 — APPROVED** [review]

**Review by kalyankrishna1 — APPROVED** [review]

**Pull request merged**